starlite/lifecycle_hooks.py

```diff
@@ -0,0 +1,73 @@
+from functools import partial
+from typing import TYPE_CHECKING, Any, Callable, Optional, Type, TypeVar
+
+from anyio.to_thread import run_sync
+from typing_extensions import Literal
+
+from starlite.utils import ensure_unbound, is_async_callable
+
+if TYPE_CHECKING:
+    from starlite.handlers.http import HTTPRouteHandler
+
+T_hook = TypeVar("T_hook", bound="LifecycleHook")
+
+
+class LifecycleHook:
+    """
+    Container for assignment of lifecycle hook handlers to instances.
+
+    A callable assigned to a class is implicitly converted to a bound method on an instance:
+
+        >>> def a_callable(): ...
+        ...
+        >>> class C:
+        ...   callable = a_callable
+        ...
+        >>> C().callable()
+        Traceback (most recent call last):
+          File "<stdin>", line 1, in <module>
+        TypeError: a_callable() takes 0 positional arguments but 1 was given
+
+    `LifecycleHook` supports the pattern of storing handlers as attributes of the application layers, and caching
+    resolved handlers on `HTTPRouteHandler` instances.
+
+    Additionally, it pre-computes whether the handler function is async, simplifying calling the handler during the
+    request/response cycle.
+    """
+
+    def __init__(self, hook: Callable[..., Any]) -> None:
+        if is_async_callable(hook):
+            self.hook = hook
+        else:
+            self.hook = partial(run_sync, hook)
+
+    async def __call__(self, *args: Any) -> Any:
+        return await self.hook(*args)
+
+    @classmethod
+    def resolve_for_handler(
+        cls: Type[T_hook],
+        route_handler: "HTTPRouteHandler",
+        attribute_key: Literal["after_request", "after_response", "before_request"],
+    ) -> Optional[T_hook]:
+        """
+        Resolves `attribute_key` for `route_handler`.
+
+        If a hook is registered, returns an instance of `LifecycleHook`, otherwise returns `None`.
+
+        Args:
+            route_handler (HTTPRouteHandler): the handler to have the hook resolved.
+            attribute_key (Literal["after_request", "after_response", "before_request"]): hook layer attribute name.
+
+        Returns:
+            HTTPRouteHandler | None
+        """
+        hook = None
+        for layer in route_handler.ownership_layers:
+            layer_hook = getattr(layer, attribute_key, None)
+            if layer_hook:
+                hook = ensure_unbound(layer_hook)
```

Comment:
See #284 - I think we should limit the unbinding to within `Controller.__init__()` so that we can support instance methods and callable classes as hooks.